src/AggregateFunctions/AggregateFunctionDistinctJSONPaths.cpp

```diff
@@ -0,0 +1,350 @@
+        for (size_t i = start; i != end; ++i)
+            data.insert(shared_data_paths->getDataAt(i).toString());
+    }
+
+    void addWholeColumn(const ColumnObject & column, const std::unordered_map<String, String> &)
+    {
+        for (const auto & [path, _] : column.getTypedPaths())
+            data.insert(path);
+        for (const auto & [path, dynamic_column] : column.getDynamicPathsPtrs())
+        {
+            /// Add dynamic path only if it has at least one non-null value.
+            /// getNumberOfDefaultRows for Dynamic column is O(1).
+            if (dynamic_column->getNumberOfDefaultRows() != dynamic_column->size())
+                data.insert(path);
+        }
+
+        /// Iterate over all paths in shared data.
+        const auto [shared_data_paths, _] = column.getSharedDataPathsAndValues();
+        for (size_t i = 0; i != shared_data_paths->size(); ++i)
+            data.insert(shared_data_paths->getDataAt(i).toString());
+    }
+
+    void merge(const AggregateFunctionDistinctJSONPathsData & other)
+    {
+        data.insert(other.data.begin(), other.data.end());
+    }
+
+    void serialize(WriteBuffer & buf) const
+    {
+        writeVarUInt(data.size(), buf);
+        for (const auto & path : data)
+            writeStringBinary(path, buf);
+    }
+
+    void deserialize(ReadBuffer & buf)
+    {
+        size_t size;
+        readVarUInt(size, buf);
+        if (size > DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE)
+            throw Exception(ErrorCodes::TOO_LARGE_ARRAY_SIZE, "Too large array size (maximum: {}): {}", DISTINCT_JSON_PATHS_MAX_ARRAY_SIZE, size);
+
+        String path;
+        for (size_t i = 0; i != size; ++i)
+        {
+            readStringBinary(path, buf);
+            data.insert(path);
+        }
+    }
+
+    void insertResultInto(IColumn & column)
+    {
+        /// Insert paths in sorted order for better output.
+        auto & array_column = assert_cast<ColumnArray &>(column);
+        auto & string_column = assert_cast<ColumnString &>(array_column.getData());
+        std::vector<String> sorted_data(data.begin(), data.end());
+        std::sort(sorted_data.begin(), sorted_data.end());
+        for (const auto & path : sorted_data)
+            string_column.insertData(path.data(), path.size());
+        array_column.getOffsets().push_back(string_column.size());
+    }
+
+    static DataTypePtr getResultType()
+    {
+        return std::make_shared<DataTypeArray>(std::make_shared<DataTypeString>());
+    }
+};
+
+struct AggregateFunctionDistinctJSONPathsAndTypesData
+{
+    static constexpr auto name = "distinctJSONPathsAndTypes";
+
+    std::unordered_map<String, std::unordered_set<String>> data;
+
+    void add(const ColumnObject & column, size_t row_num, const std::unordered_map<String, String> & typed_paths_type_names)
+    {
+        for (const auto & [path, _] : column.getTypedPaths())
+            data[path].insert(typed_paths_type_names.at(path));
```

Comment:
Based on this and the similar code in `AggregateFunctionDistinctJSONPathsData` I assume we list the typed paths in both cases regardless if they contain any value or not. If I am right, then it is kind of surprising for me, so maybe explicitly mention this in the docs.

Reply:
Yes. It's because typed paths are not `Dynamic`, they have user specified types (so they can be even not Nullable) and we cannot know if the row contains the value from the data or it's just a default value.
We treat typed paths as they have values in each row

Reply:
I will add it to the docs